buf.yaml

```
@@ -30,3 +30,6 @@ lint:
 breaking:
   use:
     - WIRE_JSON
+    - FIELD_NO_DELETE
+  except:
+    - FIELD_SAME_ONEOF
```

```suggestion
    - FILE
```
@miklos-da @fabiotudone-da @moritzkiefer-da : reading `https://docs.buf.build/breaking/rules/`, I'd expect that we want to use `FILE` level compatibility for:
* daml-lf/archive/src/main/protobuf
* daml-lf/transaction/src/main/protobuf
* ledger-api/grpc-definitions
* ledger/ledger-configuration/protobuf
as they all can be referenced outside of the repo by third parties.

And we want to use at least `WIRE_JSON` with
* ledger/participant-integration-api/src/main/protobuf
* ledger/participant-state/kvutils/src/main/protobuf

Albeit `FILE` will work fine for `ledger/participant-integration-api/src/main/protobuf`

Questions:
*  @miklos-da @fabiotudone-da : would `FILE` work as well for `kvutils`?
*  @moritzkiefer-da : are you OK with `FILE` for the `daml-lf` protos?

---

Looking at this I didn’t realize this was enabled for daml-lf archive protobufs and looking at the current definition I’m not sure it’s correct. You are only checking Daml-LF dev which is explicitly the version that we can and do break so I don’t think the current definition makes sense. I think the check for daml-lf that we want is different, we want to check that dev is backwards compatible to the stable LF releases (those are immutable, we check hashes) not to the dev version included in the latest SDK release. So my suggestion there would be to disable it for daml-lf/archive as part of this PR and then we (doesn’t have to be @fabiotudone-da) can add a check for that separately.

For daml-lf/transaction I’d go with `WIRE` or `WIRE_JSON` (we don’t care about JSON encoding afaik so wire would be sufficient). These are not public APIs, as long as we can read old stuff that’s fine, I don’t see why we should force us to never break generated code if we control all of it.

---

Thanks for the feedback @cocreature . 

@fabiotudone-da : reading https://docs.buf.build/breaking/configuration it looks like the right option for per-proto-group configurations is to place a `buf.yaml` at their roots instead of having a single `buf.yaml` at the root of the repository. Do you understand it the same way?

If yes, could you create the ones for
* `ledger-api/grpc-definitions` using `FILE`
* `ledger/ledger-configuration/protobuf` using `FILE`
* `ledger/participant-integration-api/src/main/protobuf` using `FILE`
* `ledger/participant-state/kvutils/src/main/protobuf` using your team's configuration of choice

---

@meiersi-da:

> we want to use FILE level compatibility for
> ...
> ledger-api/grpc-definitions

Do we offer source-level compatibility for the gRPC API?

> ledger/ledger-configuration/protobuf

It looks like we persist it but we do not offer API access to it, so why is source-level compatibility important?

> would FILE work as well for kvutils?

If I remember correctly we had trouble with it but @miklos-da may remember better. Are you asking in order to try and consolidate the checks category or do you see a reason why source-level compatibility can be important here?

---

@fabiotudone-da : reading https://docs.buf.build/breaking/configuration it looks like the right option for per-proto-group configurations is to place a `buf.yaml` at their roots instead of having a single `buf.yaml` at the root of the repository. Do you understand the docs the same way?

If yes, could you create the ones for
* `daml-lf/transaction` using `WIRE_JSON`
* `ledger-api/grpc-definitions` using `FILE`
* `ledger/ledger-configuration/protobuf` using `FILE` as it is included in the `gRPC` services
* `ledger/participant-integration-api/src/main/protobuf` using `WIRE_JSON`
* `ledger/participant-state/kvutils/src/main/protobuf` using `WIRE_JSON`

This should also take care of @cocreature's request to not check the `daml-lf/archive/dev` versions.

---

> @meiersi-da:
> 
> > we want to use FILE level compatibility for
> > ...
> > ledger-api/grpc-definitions

Our docs are not explicit on this, but it's what I would aim for, as we want to make it as easy as possible to build applications against our API.

> Do we offer source-level compatibility for the gRPC API?
> 
> > ledger/ledger-configuration/protobuf
> 
> It looks like we persist it but we do not offer API access to it, so why is source-level compatibility important?

It is referenced from the gRPC API

---

> > > ledger/ledger-configuration/protobuf
> > 
> > 
> > It looks like we persist it but we do not offer API access to it, so why is source-level compatibility important?
> 
> It is referenced from the gRPC API

The only import of `ledger_configuration.proto` I could find is in `daml_kvutils.proto`. Am I missing something?

---

@fabiotudone-da : it looks like I was mistaken wrt its inclusion in the gRPC Ledger API. It is though present in several migrations for the IndexDB, as the `configuration` column in the `configuration_entries` table is encoded using that protobuf. See https://github.com/digital-asset/daml/search?q=ledger_configuration&type=code

---

So using `WIRE_JSON` for `ledger/ledger-configuration/protobuf` seems like the right option.

---

Why not `WIRE`?

---

I think it was mostly for consistency, i.e. it can be confusing for the reader to see changing field names. But I don't recall precisely. It was introduced [here](https://github.com/digital-asset/daml/pull/10411/files#diff-1a5ba9cba93e971f532139f694d7da802776bfe578e3f753b9c3f25968dbf42dR27) so @miklos-da may have a better answer.

---

As we're providing bindings, I'm not sure source-level compatibility is required for the Ledger API (hence, `WIRE` should suffice).
For `ledger/ledger-configuration/protobuf` `WIRE` should be also enough as it is only being used internally by kvutils (hence, we don't care about changing field names).

---

@meiersi-da Yes, I'm aware that we are importing `com/daml/daml_lf_dev/daml_lf.proto` in `daml_kvutils.proto`, this was required for embedding Daml-LF data and improving compression efficiency (we could replace relevant Daml-LF message references with an `Any` message type for more flexibility). I agree @cocreature's team should take care of putting the right backwards compatibility checks in place that works for their proto files.

---

> As we're providing bindings, I'm not sure source-level compatibility is required for the Ledger API (hence, WIRE should suffice).

The Ledger gRPC API is part of public documentation, so I think it's more important there to have stability for field names; for this reason I propose to keep `WIRE_JSON`.

We could also try and upgrade to `FILE` if we decide to strive for source-level compatibility, for cases (present or potential, not sure) when an user wants to bypass bindings and still upgrade easily, but I'd prefer we take a conscious and shared decision about this at a product reasoning level (and then also make it explicit in public docs).

---

abc701 is the best I could do. Notes:

-  Without us publishing `buf` modules at this stage (so that we can declare dependencies between them), we must explicitly list all roots containing referenced `.proto` files and parent paths cannot be referenced, so I had to move `buf-kvutils.yaml` up to the root. For consistency, the other `buf` configuration files are at the root level too, prefixed with `buf-` so at least they show up close to each other.

- Using the `buf` registry to retrieve Google's `status.proto` requires the dependent to also adhere to `buf` module packaging (in particular, dedicated root with a single configuration that must be named `buf.yaml`) and, since we don't do that due to the previous point, I had to commit `status.proto`.

- On the other hand, this structure has the advantage that the check script doesn't have to know anything about the directory structure.

Improvements are possible but I'd leave them for further PRs to avoid scope creep, especially considering that they may end up being possible only after reorganizing the code and/or reviewing dependencies between our protos.

@meiersi-da @miklos-da I'd like especially your feedback (and approval, if you agree) about two main things in this context:

- Whether you see serious issues with this (transitional) `buf` structure for the purpose of this PR.
- Whether we can agree on the modules split and the checks for the purpose of this PR.

---

Also note that I had to dismantle at this stage the previous manual `fmt.sh`-based protobuf checks.

---

@fabiotudone-da This sounds great to me, we can always fine-tune the rules later!

---

Resolving this thread because further feedback is being addressed in new threads.